---
title: "CRC datasets heatmap"
output:
  html_document:
    df_print: paged
---
```{r include=FALSE}
knitr::opts_chunk$set(echo = FALSE, fig.path = 'CRC_heatmap_plots/', dev = c('png','svg'), warning = FALSE)
library(tidyverse)
library(magrittr)
library(ComplexHeatmap)
library(circlize)
library(curatedMetagenomicData)
MTA_palette <- c("#0039A6","#FF6319","#6CBE45","#996633","#A7A9AC","#FCCC0A","#808183","#EE352E","#00933C","#B933AD")
CRC_datasets <- strsplit(read_file("CRC_datasets.lst"), '\n')[[1]]
ThomasA_2019_datasets <- c('FengQ_2015', 'YuJ_2015', 'ZellerG_2014', 'VogtmannE_2016', 'HanniganGD_2017', 'ThomasAM_2018a', 'ThomasAM_2018b')
```


```{r load_data}
mpa_table <- read_tsv("/shares/CIBIO-Storage/CM/news/users/paolo.manghi/ongoing_paper/MetaPhlAn3/FIG2/MetaPhlAn3.0/qstat_02/METAPHLAN3_CRC_DATA.tsv", col_names = FALSE, skip = 42)
colnames(mpa_table)[1] <- 'species'

mpa_table$species <- gsub('^.*s__','',mpa_table$species)

metadata <- read_tsv("/shares/CIBIO-Storage/CM/news/users/paolo.manghi/ongoing_paper/MetaPhlAn3/FIG2/MetaPhlAn3.0/qstat_02/METAPHLAN3_CRC_DATA.tsv", col_names = FALSE, n_max = 42) %>% 
  t() %>% 
  as.data.frame(row.names = FALSE, stringsAsFactors=FALSE)
colnames(metadata) <- as.character(metadata[1,])
metadata <- metadata[-1,] 
colnames(mpa_table)[-1] <- metadata$sampleID

markers <- c("Roseburia_intestinalis", "Roseburia_sp_CAG_303", "Streptococcus_oralis", "Bacteroides_nordii", "Intestinimonas_butyriciproducens", "Prevotella_intermedia", "Fretibacterium_fastidiosum", "Methanobrevibacter_smithii", "Bacteroides_fragilis", "Mogibacterium_diversum", "Porphyromonas_asaccharolytica", "Eisenbergiella_tayi", "Bulleidia_extructa", "Peptostreptococcus_anaerobius", "Solobacterium_moorei", "Peptostreptococcus_stomatis", "Gemella_morbillorum", "Fusobacterium_nucleatum", "Dialister_pneumosintes", "Parvimonas_micra")
```


# Table 1
```{r}
filt_metadata <- metadata %>% 
                  select(dataset_name, study_condition, country, s__age, gender, BMI) %>% 
                  mutate(BMI = as.numeric(BMI),
                         age = as.numeric(s__age),
                         study_condition = factor(study_condition),
                         gender = factor(gender),
                         country = factor(country))
filt_metadata %>% group_by(dataset_name, study_condition) %>% summarise(n())
```


```{r}
mpa_table_markers <-
  mpa_table %>%
  filter(species %in% markers) %>%
  arrange(desc(rowMeans(.[,-1]))) %>% 
  select(species, noquote(metadata %>% arrange(study_condition) %>% select(sampleID)) %>% t %>% as.character) %>% 
  mutate_if(is.numeric, funs(log10(1+.))) %>% 
  mutate(species = gsub('_',' ', species)) %>% 
  column_to_rownames('species')
```

# Heatmap: Abundance of CRC Biomarkers
```{r}
annot_v <- metadata %>% 
  arrange(study_condition) %>%  
  select(sampleID, Condition=study_condition, Dataset=dataset_name) %>% 
  mutate(Condition =  sub("(.)", "\\U\\1", Condition, perl=TRUE)) %>% 
  column_to_rownames("sampleID")

cols <- list("Condition" = c("Control" = "darkolivegreen3", "CRC" = "orange"),
             "Dataset" = setNames(MTA_palette, sort(unique(annot_v$Dataset)))
            )

# hclust_controls <- mpa_table_markers %>% 
#   select(annot_v %>% rownames_to_column('sampleID') %>% filter(Condition == 'Control') %>% select(sampleID) %>% t %>% as.character) %>% 
#   t %>% 
#   dist(method = "euclidean") %>% 
#   hclust(method = "average")
# 
# hclust_CRC <- mpa_table_markers %>% 
#   select(annot_v %>% rownames_to_column('sampleID') %>% filter(Condition == 'CRC') %>% select(sampleID) %>% t %>% as.character) %>% 
#   t %>% 
#   dist(method = "euclidean") %>% 
#   hclust(method = "average")
```

```{r metaphlan3_CRC_biomarkers, fig.width=10, fig.height=5}
x <- colorRamp2(breaks = seq(0,1,0.1), colors = hcl.colors(11, palette = 'Inferno'), transparency = 0.1, )
attributes(x)$breaks <- seq(0,2,0.1)
Heatmap(as.matrix(mpa_table_markers),
        top_annotation = HeatmapAnnotation(df = annot_v, col = cols),
        col = x,
        name = "Relative abundance(Log))",
        column_split = annot_v$Condition,
        column_gap = unit(1,"points"),
        column_title = rep("",2),
        cluster_rows = FALSE,
        show_column_names = FALSE,
        cluster_column_slices = TRUE,
        show_column_dend = FALSE,
        row_names_gp = gpar(fontface = 'italic', fontfamily = "sans"),
        heatmap_legend_param = list(direction = 'horizontal', border = 'black', position = 'top', labels = c('0', '0.25', '0.5', '0.75', '1'))
)

```

# Heatmap: Relative abundance of top 30 species
```{r}
top30_mpa_table <- mpa_table %>%
  arrange(desc(rowMeans(.[,-1]))) %>% 
  select(species, noquote(metadata %>% arrange(study_condition) %>% select(sampleID)) %>% t %>% as.character) %>% 
  mutate_if(is.numeric, funs(log10(1+.))) %>% 
  mutate(species = gsub('_',' ', species)) %>% 
  top_n(10) %>% 
  column_to_rownames('species')


annot_v <- metadata %>% 
  arrange(study_condition) %>%  
  select(sampleID, Condition=study_condition, Dataset=dataset_name) %>% 
  mutate(Condition =  sub("(.)", "\\U\\1", Condition, perl=TRUE)) %>% 
  column_to_rownames("sampleID")

cols <- list("Condition" = c("Control" = "darkolivegreen3", "CRC" = "orange"),
             "Dataset" = setNames(MTA_palette, sort(unique(annot_v$Dataset)))
            )
```

```{r metaphlan3_CRC_top30, fig.width=10, fig.height=5}
x <- colorRamp2(breaks = seq(0,1,0.1), colors = hcl.colors(11, palette = 'Inferno'), transparency = 0.1, )
attributes(x)$breaks <- seq(0,2,0.1)
Heatmap(as.matrix(top30_mpa_table), 
        top_annotation = HeatmapAnnotation(df = annot_v, col = cols),
        col = x,
        name = "Relative abundance(Log))",
        column_split = annot_v$Condition,
        column_gap = unit(1,"points"),
        column_title = rep("",2),
        cluster_rows = FALSE,
        show_column_names = FALSE,
        cluster_column_slices = TRUE,
        show_column_dend = FALSE,
        row_names_gp = gpar(fontface = 'italic', fontfamily = "sans"),
        heatmap_legend_param = list(direction = 'horizontal', border = 'black', position = 'top', labels = c('0', '0.25', '0.5', '0.75', '1'))
)

```
## HUMAnN 3
### Normalize per CPM
```{bash}
conda activate humann-3.0
for dataset in FengQ_2015 GuptaA_2019 ThomasAM_2019_a ThomasAM_2019_b ThomasAM_2019_c VogtmannE_2016 WirbelJ_2018 YachidaS_2019 YuJ_2015 ZellerG_2014
do
  find /shares/CIBIO-Storage/CM/scratch/data/meta/${dataset}/humann-3.0_v30_CHOCOPhlAn_201901 -name *_genefamilies.tsv | xargs -I {} -P 50 bash -c 'x={}; humann2_renorm_table --input ${x} --units cpm --update-snames 2>/dev/null --output ${x/genefamilies/genefamilies-cpm}'
done

for dataset in FengQ_2015 GuptaA_2019 ThomasAM_2019_a ThomasAM_2019_b ThomasAM_2019_c VogtmannE_2016 WirbelJ_2018 YachidaS_2019 YuJ_2015 ZellerG_2014
do
  find /shares/CIBIO-Storage/CM/scratch/data/meta/${dataset}/humann-3.0_v30_CHOCOPhlAn_201901 -name *_genefamilies.tsv | xargs -I {} -P 50 bash -c 'x={}; humann2_renorm_table --input ${x} --units relab --update-snames 2>/dev/null --output ${x/genefamilies/genefamilies-relab}'
done
```


```{r load_humann_pwy, warning=FALSE, message=FALSE}
CRC_datasets <- c("FengQ_2015", "GuptaA_2019", "ThomasAM_2019_a", "ThomasAM_2019_b", "ThomasAM_2019_c", "VogtmannE_2016", "WirbelJ_2018", "YachidaS_2019", "YuJ_2015", "ZellerG_2014")
hnn_pathabundance <- lapply(CRC_datasets, function(dataset) { 
  dataset_hnn_path <- paste('/shares/CIBIO-Storage/CM/scratch/data/meta', dataset, 'humann-3.0_v30_CHOCOPhlAn_201901', sep = '/')
  lapply(list.files(dataset_hnn_path), function(sample) {
    if(sample %in% metadata$sampleID){
      paste(dataset_hnn_path, sample, paste0(sample,"_pathabundance.tsv"), sep = '/')
    }
  })
}) %>% unlist

hnn_normalizedgenefamilies <- lapply(CRC_datasets, function(dataset) {
  dataset_hnn_path <- paste('/shares/CIBIO-Storage/CM/scratch/data/meta', dataset, 'humann-3.0_v30_CHOCOPhlAn_201901', sep = '/')
  lapply(list.files(dataset_hnn_path), function(sample) {
    if(sample %in% metadata$sampleID){
      paste(dataset_hnn_path, sample, paste0(sample,"_genefamilies-cpm.tsv"), sep = '/')
    }
  })
}) %>% unlist

hnn_pathcoverage <- lapply(CRC_datasets, function(dataset) {
  dataset_hnn_path <- paste('/shares/CIBIO-Storage/CM/scratch/data/meta', dataset, 'humann-3.0_v30_CHOCOPhlAn_201901', sep = '/')
  lapply(list.files(dataset_hnn_path), function(sample) {
    if(sample %in% metadata$sampleID){
    paste(dataset_hnn_path, sample, paste0(sample,"_pathcoverage.tsv"), sep = '/')
    }
  })
}) %>% unlist

cutC_NR90_ids <- strsplit(read_file("cutC_NR90"), '\n')[[1]]
cutD_NR90_ids <- strsplit(read_file("cutD_NR90"), '\n')[[1]]

hnn_pwyab <- lapply(hnn_pathabundance, function(sample_path) {
                          if (file.exists(sample_path)){
                            read_tsv(sample_path) %>% 
                              pivot_longer(-1, names_to = 'sample', values_to = 'relative_abundance') %>% 
                              filter(str_detect(`# Pathway`, '\\|', negate=TRUE)) %>% 
                              mutate(sample = str_replace(sample, '_Abundance',''))
                          }
                  }) %>% 
              bind_rows %>% 
              pivot_wider(names_from = 'sample', values_from = 'relative_abundance')
  


# hnn_genfam_cutC <- lapply(hnn_genefamilies, function(sample_path) {
#                           if (file.exists(sample_path)){
#                             read_tsv(sample_path) %>%
#                               filter(str_detect(.$`# Gene Family`, paste(cutC_NR90,collapse = '|'))) %>% 
#                               pivot_longer(-1, names_to = 'sample', values_to = 'RPK') %>%
#                               mutate(sample = str_replace(sample, '_Abundance-RPKs',''))
#                           }
#                   }) %>%
#               bind_rows %>%
#               pivot_wider(names_from = 'sample', values_from = 'RPK')

hnn_pwycov <- lapply(hnn_pathcoverage, function(sample_path) {
                          if (file.exists(sample_path)){
                            read_tsv(sample_path) %>%
                              pivot_longer(-1, names_to = 'sample', values_to = 'coverage') %>%
                              mutate(sample = str_replace(sample, '_Coverage',''))
                          }
                  }) %>%
              bind_rows %>%
              pivot_wider(names_from = 'sample', values_from = 'coverage')
```

```{bash eval=FALSE, include=FALSE}
for dataset in FengQ_2015 GuptaA_2019 ThomasAM_2019_a ThomasAM_2019_b ThomasAM_2019_c VogtmannE_2016 WirbelJ_2018 YachidaS_2019 YuJ_2015 ZellerG_2014
do
  find /shares/CIBIO-Storage/CM/scratch/data/meta/${dataset}/humann-3.0_v30_CHOCOPhlAn_201901 -name *_genefamilies-cpm.tsv | xargs -I {} -P 100 bash -c 'x={}; grep -Fwf /shares/CIBIO-Storage/CM/news/users/francesco.beghini/chocophlan_paper/CRC_metaphlan3/cutC_NR90 $x | grep -v "|" | awk -v a=${x##*/} "{print \$0 \"\t\" a }"'
done > merged_cutC_RPK.tsv

for dataset in FengQ_2015 GuptaA_2019 ThomasAM_2019_a ThomasAM_2019_b ThomasAM_2019_c VogtmannE_2016 WirbelJ_2018 YachidaS_2019 YuJ_2015 ZellerG_2014
do
  find /shares/CIBIO-Storage/CM/scratch/data/meta/${dataset}/humann-3.0_v30_CHOCOPhlAn_201901 -name "*_genefamilies-relab.tsv" | xargs -I {} -P 100 bash -c 'x={}; grep -Fwf /shares/CIBIO-Storage/CM/news/users/francesco.beghini/chocophlan_paper/CRC_metaphlan3/cutC_NR90 $x | grep -v "|" | awk -v a=${x##*/} "{print \$0 \"\t\" a }"'
done > merged_cutC_RPK_relabnorm.tsv
```

```{r load_cutC_RPK, message=FALSE, warning=FALSE}
NR90_cutC <- read_tsv("merged_cutCD_RPK_CPM.tsv", col_names = c('NR90', 'RPK', 'sample')) %>% 
  filter(NR90 %in% cutC_NR90_ids) %>% 
  mutate(sample = str_remove(sample, "_genefamilies.*")) %>% 
  filter(sample %in% metadata$sampleID) %>% 
  pivot_wider(names_from = 'sample', values_from = 'RPK', values_fill = list(RPK = 0)) %>% 
  column_to_rownames('NR90')

missing_samples_lst <- metadata[!(metadata$sampleID %in% colnames(NR90_cutC)),2]
missing_samples <- as.data.frame(matrix(data = 0, nrow = nrow(NR90_cutC), ncol = length(missing_samples_lst)))
colnames(missing_samples) <- missing_samples_lst
rownames(missing_samples) <- rownames(NR90_cutC)

## cutC + cutD // pathway PWY-7167

NR90_cutC <- cbind(missing_samples, NR90_cutC)
```

```{r load_cutD_RPK, message=FALSE, warning=FALSE}
NR90_cutD <- read_tsv("merged_cutCD_RPK_CPM.tsv", col_names = c('NR90', 'RPK', 'sample')) %>% 
  filter(NR90 %in% cutD_NR90_ids) %>% 
  mutate(sample = str_remove(sample, "_genefamilies.*")) %>% 
  filter(sample %in% metadata$sampleID) %>% 
  pivot_wider(names_from = 'sample', values_from = 'RPK', values_fill = list(RPK = 0)) %>% 
  column_to_rownames('NR90')

missing_samples_lst <- metadata[!(metadata$sampleID %in% colnames(NR90_cutD)),2]
missing_samples <- as.data.frame(matrix(data = 0, nrow = nrow(NR90_cutD), ncol = length(missing_samples_lst)))
colnames(missing_samples) <- missing_samples_lst
rownames(missing_samples) <- rownames(NR90_cutD)

## cutC + cutD // pathway PWY-7167 L-carnitine (PWY-3641). Maybe checkout yeaW 

NR90_cutD <- cbind(missing_samples, NR90_cutD)
```


```{r, fig.height=8, fig.width=10}
x <- colorRamp2(breaks = seq(0,1,0.1), colors = hcl.colors(11, palette = 'Inferno'), transparency = 0.1, )
attributes(x)$breaks <- seq(0,2,0.1)
Heatmap(as.matrix(log10(1+NR90_cutC)), 
        top_annotation = HeatmapAnnotation(df = annot_v[colnames(NR90_cutC),], col = cols),
        name = "RPK",
        column_split = annot_v[colnames(NR90_cutC),]$Condition,
        column_gap = unit(1,"points"),
        # column_title = rep("",2),
        # cluster_rows = FALSE,
        show_column_names = FALSE,
        cluster_column_slices = TRUE,
        show_column_dend = FALSE,
        row_names_gp = gpar(fontface = 'italic', fontfamily = "sans"),
        heatmap_legend_param = list(direction = 'horizontal', border = 'black', position = 'top', labels = c('0', '0.25', '0.5', '0.75', '1'))
)
```

```{r}
NR90_cutC %>%
  rownames_to_column('NR90') %>%
  pivot_longer(-NR90, names_to = 'sample', values_to = 'RPK') %>%
  full_join(metadata, by = c('sample' = 'sampleID')) %>% 
  mutate(study_condition = factor(study_condition)) %>% 
  mutate(dataset_name = str_replace(dataset_name, 'ThomasAM_2019_c', 'YachidaS_2019')) %>%
  group_by(dataset_name) %>%
  summarise(wilcox = wilcox.test(RPK ~ study_condition, paired=FALSE)$p.value %>% round(4))
```

```{r}
NR90_cutD %>%
  rownames_to_column('NR90') %>%
  pivot_longer(-NR90, names_to = 'sample', values_to = 'RPK') %>%
  full_join(metadata, by = c('sample' = 'sampleID')) %>% 
  mutate(study_condition = factor(study_condition)) %>% 
  mutate(dataset_name = str_replace(dataset_name, 'ThomasAM_2019_c', 'YachidaS_2019')) %>%
  group_by(dataset_name) %>%
  summarise(wilcox = wilcox.test(RPK ~ study_condition, paired=FALSE)$p.value %>% round(10))
```

```{r}
NR90_cutC %>%
  rownames_to_column('NR90') %>%
  pivot_longer(-NR90, names_to = 'sample', values_to = 'RPK') %>%
  full_join(metadata, by = c('sample' = 'sampleID')) %>% 
  mutate(RPK = (1+RPK)) %>% 
  ggplot() +
  geom_boxplot(aes(study_condition,log10(RPK))) +
  facet_wrap(~dataset_name) +
  scale_y_log10()
```


# Heatmap: Top 30 pathways
```{r message=FALSE, warning=FALSE}
hnn_pwyab %>% 
  separate(col = 1, into = c('pwy' ,'org'), sep = '\\|' ) %>% 
  filter(is.na(org)) %>% 
  arrange(desc(rowMeans(.[,-c(1,2)]))) %>% 
  top_n(30) %>% 
  select(1) %>% 
  kableExtra::kable()
```



```{r}
setExperimentHubOption('CACHE', '/shares/CIBIO-Storage/CM/news/tmp/francesco.beghini/')
x <- curatedMetagenomicData(sapply(ThomasA_2019_datasets, function(x) paste0(x, '*pathabundance*')), dryrun = FALSE)
x <- mergeData(x)

all_pwy <- exprs(x) %>%
  as.data.frame() %>% 
  rownames_to_column('pathway') %>% 
  separate(col = 1, into = c('pwy' ,'org'), sep = '\\|' ) %>% 
  filter(is.na(org)) %>% 
  select(-org) %>% 
  arrange(desc(rowMeans(.[,-1]))) %>% 
  top_n(30) %>% 
  select(1)

```

